src/rpc/client.cpp

```diff
@@ -47,8 +50,28 @@ UniValue CallRPC(const string& strMethod, const UniValue& params)
     asio::ssl::stream<asio::ip::tcp::socket> sslStream(io_context, context);
     SSLIOStreamDevice<asio::ip::tcp> d(sslStream, fUseSSL);
     iostreams::stream< SSLIOStreamDevice<asio::ip::tcp> > stream(d);
-    if (!d.connect(gArgs.GetArg("-rpcconnect", "127.0.0.1"), gArgs.GetArg("-rpcport", ToString(GetDefaultRPCPort()))))
-        throw runtime_error("couldn't connect to server");
+
+    bool fWait = gArgs.GetBoolArg("-rpcwait", false); // -rpcwait means try until server has started or timeout is reached
+    int timeout = gArgs.GetArg("-rpcwaittimeout", DEFAULT_WAIT_CLIENT_TIMEOUT); // The max time to wait
+
+    std::chrono::seconds deadline =  GetTime<std::chrono::seconds>() + std::chrono::seconds{timeout};
+
+    do {
+        // If connection succeeds, immediately break. No need to wait.
+        if (d.connect(gArgs.GetArg("-rpcconnect", "127.0.0.1"),
+                      gArgs.GetArg("-rpcport", ToString(GetDefaultRPCPort())))) {
+            break;
+        }
+
+        std::chrono::seconds now = GetTime<std::chrono::seconds>();
+
+        // Note that timeout <= 0 means wait until connected with one second between connection attempts.
+        if (fWait && (timeout <= 0 || now < deadline)) {
+            UninterruptibleSleep(std::chrono::seconds{1});
+        } else {
+            throw runtime_error("couldn't connect to server");
+        }
+    } while (fWait);
```

Comment:
nit: Extremely minor optimization.
```suggestion
    } while (true);
```

Reply:
I am going to leave it the way it was on this one. If fWait is false, in the original, the sleep is turned off because fWait is false, and then the do... while exits because the while condition is false, which is equivalent to the original code prior to this PR. If you change the do .. while condition to true, then if fWait is false, the while loop will actually attempt to iterate again if the first connection attempt fails, with no sleep, and this will continue indefinitely until connected.

Reply:
The while loop cannot continue as runtime_error will be thrown if fWait is false.

Reply:
Hmm.. You're right.